src/merge.c

```diff
@@ -3239,6 +3239,10 @@ int git_merge(
 
 	assert(repo && their_heads);
 
+	if((error = git_repository_index(&index, repo) < 0) ||
```

Comment:
Actually, I think that the error message set by `git_repository_index` itself would be superior to changing the error message.  It will have the OS error details, like `permission denied`, which we would not want to overwrite with a less descriptive error.